random methods for probability distributions

mwibrow opened the pull request:
This is a **not ready to merge** but a WIP which relates to the pull request #442. 

I have implemented pdf, cdf and random methods for most of the continuous distributions. They are mostly wrappers around scipy/numpy stuff and I've only done tests for the majority of the pdfs. I have incorporated the `draw_values` idea of @fonnesbeck to fix parameter values, which I _think_ works, so far.

Commit:
Added pdf, cdf and random methods to continuous distributions.

Removed 0 from Rplus and added RZPlus to be old Rplus.

RNZplus is "Non-Zero" Rplus. Rplus once again contains zero

Added support function to dist_math.
Like bound but for numpy and returns zero if conditions not met.

Added tests for some pdfs.

Corrected mu/sd paratrization for beta distribution.
Required changing of test to convert scipy nan to -inf.

Made all domains contain floats.

Added tests for t, cauchy, half-cauchy and gamma pdfs

Fixed pickling error with continuous distributions.

Changed to builtin getattr and setattr where appropriate.

Add more getattr.

Added parametrize method to distributions with alternative parametrizations.

Fixed raised error on class declaration (!)

Added pdf tests for more continuous distributions.

Add more pdf tests.

Added tests for ex-gaussian and inverse gaussian pdfs.

Added pdf tests for GeneralizedExtremeValue distribution.

Fixex typos in doc strings and comments.

twiecki commented on pymc3/distributions/continuous.py:
```diff
@@ -26,6 +36,82 @@ class UnitContinuous(Continuous):
+    return nodes
+
+
+"""
```
Can you move that to be a docstring?

twiecki commented on pymc3/distributions/continuous.py:
```diff
@@ -141,22 +286,85 @@ class Normal(Continuous):
         self.variance = 1. / self.tau
 
-    def logp(self, value):
+    def parametrize(self, mu=None, tau=None, sd=None):
```
I like this refactoring.

twiecki commented on pymc3/distributions/continuous.py:
```diff
@@ -141,22 +286,85 @@ class Normal(Continuous):
+                Parameter values to fix.                
+        """
+        mu, tau = draw_values([self.mu, self.tau], point=point)                                     
+        return np.sqrt(0.5 * tau / pi) * np.exp(-0.5 * tau * (x - mu) ** 2) 
```
Any reason this is not theano? I think it's nice to have but I'm not sure where the PDF would be useful specifically. If it were to be used in modeling it would have to be Theano.

mwibrow replied:
It part, the `pdf`/`cdf`/`random` methods are numpy/scipy based as ultimately (as I've started in another branch) I was heading towards a decorator which enables these methods to be called on an instance but also on the class as well (i.e., without class instantiation or model context) with specific numerical values:

``` python
y = Normal.random(mu=0., sd=2)
```

So its basically just useful for _me_ and I usually use it just for plotting or generating samples (which when done 'properly' using a model context is a bit slow on my computer), but the changes to make the `pdf` method theano based are probably trivial.

twiecki commented on pymc3/distributions/continuous.py:
```diff
@@ -141,22 +286,85 @@ class Normal(Continuous):
+            The iid x(s) to evaluate the log-likelihood. 
+        """
         tau = self.tau
-        sd = self.sd
```
Actually enforcing `sd > 0` is required because `sd` could be negative and yet `tau` would be positive.

twiecki commented on pymc3/distributions/continuous.py:
```diff
@@ -165,30 +373,89 @@ class HalfNormal(PositiveContinuous):
         )
 
-
+    
```
Please look out for stray tabs.

twiecki commented on pymc3/distributions/continuous.py:
```diff
@@ -309,35 +633,92 @@ class Beta(UnitContinuous):
         elif (mu is not None) and (sd is not None):
-            alpha = mu * sd
-            beta = (1 - mu) * sd
+            alpha = (1 - mu) * (mu ** 2) / (sd ** 2) - mu
```
Was the previous transform wrong?

mwibrow replied:
The previous transform (`alpha = mu * sd` and `beta = (1 - mu) * sd`) might actually not be wrong. But I interpreted the `mu` and `sd` parameters to be the mean and square-root of the variance, respectively, so just derived the alternative transform by substitution.

If my interpretation is correct then the previous transform was wrong. Otherwise it is me that is wrong.

twiecki commented on pymc3/distributions/continuous.py:
```diff
@@ -782,9 +1789,10 @@ def dist(*args, **kwargs):
 class ExGaussian(Continuous):    
     """
-    Exponentially modified Gaussian random variable with 
+    Expoentially modified Gaussian random variable with 
```
Typo.

twiecki commented on pymc3/distributions/dist_math.py:
```diff
@@ -41,6 +43,22 @@ def bound(logp, *conditions):
+    Parameters
+    ----------
+    x : float
+    *conditionss : booleans
```
typo

twiecki commented on pymc3/model.py:
```diff
@@ -13,6 +13,60 @@
+                        override=False):
+    """
+    Add attributes to a destination object which point to
+    methods from from a source object.
```
typo

twiecki commented on pymc3/model.py:
```diff
@@ -13,6 +13,60 @@
 __all__ = ['Model', 'Factor', 'compilef', 'fn', 'fastfn', 'modelcontext', 'Point', 'Deterministic', 'Potential']
 
 
+class InstanceMethod(object):
```
Kind of annoying we have to jump through these hoops here but I suppose this is still the cleanest way... unless we depend on `dill` instead of `pickle`.

twiecki commented on pymc3/distributions/continuous.py:
```diff
@@ -229,53 +496,109 @@ class Wald(PositiveContinuous):
     
-    def get_mu_lam_phi(self, mu, lam, phi):
+    def parametrize(self, mu, lam, phi, alpha):
+        print 'param', mu, lam, phi, alpha
```
left-over debug statement?

twiecki commented:
This is pretty awesome, many thanks for putting this up! See some comments, but nothing major. Once these are resolved, what do you think are the missing steps before we can merge this? Looks pretty good to me.

Also, I like having the PDF and CDF and I can see how they would be useful for debugging but is there another reason to include them here?

Another thought is whether we could instead compute the PDFs simply by taking `T.exp(logp)`?

Commit:
Some fixes for python3 and typos/whitespace.

Commit:
Fixed draw_values for python 2 and 3.

fonnesbeck commented:
I agree -- a herculean effort. I share the concern of including CDF and PDF methods that may only be used rarely, but must be maintained. Are there particular use cases (algorithms) you are thinking of that require CDFs on PDFs on the nominal scale?

mwibrow commented:
@twiecki it is a good point about debugging, I basically started the doing the `pdf` methods so they would be quick visual checks of the output of the `random` methods (which still need to be tested/checked).

@fonnesbeck I've got no good specific use cases. The whole thing was just something that was useful for me (actually mostly for plotting stuff) and I figured the `pdf` and `cdf` methods logically belonged in the classes. 

So, I am open to suggestions. Happy to go with the consensus and move/remove stuff as required.

twiecki commented:
So certainly I have computed the pdf (via logp) to look at it compared to the data quite a few times. Maybe that's a good enough use case. Also, we're not reimplementing the pdfs here but just forwarding to the scipy versions in most cases, so it's maintenance of the forwarding, not the math implementation. I'm still torn though. Thoughts @jsalvatier?

mwibrow commented:
Actually, if PDFs and CDFs were required, relatively maintenance-free methods could be achieved with some version of the following:

``` python
def pdf(var, x, model=None):
    model = modelcontext(model)
    return np.exp(model.fastlogp(Point({var.name:x})))

def cdf(var, x, model=None):
    model = modelcontext(model)
    s, a = 0, -inf
    _cdf = np.zeros_like(x)
    it = np.nditer(x, flags=['f_index'])
    while not it.finished:
        s += scipy.integrate.quad(lambda z: np.exp(model.fastlogp(Point({var.name:z}))), 
                                  a, it[0])[0]
        _cdf[it.index] = s
        a = it[0]
        it.iternext()
    return _cdf
```

The CDF is obviously colossally inefficient compared to native scipy but it would work for any distribution with a `logp` method. Lacks the ability fix a parameters as it stands, and would need something extra to work with transformed variables, but it is a rather large saving in code.

twiecki commented:
That's a good alternative. Let me propose this: put in the `pdf` you have above since it's so simple, leave the `random()` stuff which is the critical missing feature, and punt on the `cdf`. Thoughts?

mwibrow commented:
@twiecki Yep, that works for me. But that "simple" PDF method will end being not simple at all if it is made usable, so I reckon: kick that out as well, at least the for moment, and focus on the random sampling.

So what about testing the random methods? Other than checking the code compiles without error, presumably there is a need for goodness-of-fit testing? Kolmogorov-Smirnov gives a p-value, but if I was happy with p-values I wouldn't be doing Bayesian stuff. 

I guess I could just follow the tests from #442 for now.

Commit:
Removed code for pdfs and cdfs and pdf tests.

twiecki commented:
@mwibrow That's probably the right way forward for now, I appreciate your flexibility. And the commit is done in such a way that it can easily be reverted if we do decide to get back to it.

Regarding unittests, since we only forward to `numpy` and `scipy` I don't think we necessarily need to test the random number generation. I suppose bugs could come from wrongly transforming variables. We used KS for HDDM https://github.com/hddm-devs/hddm/blob/master/hddm/tests/test_generate.py and it worked pretty well.

jsalvatier commented on pymc3/distributions/continuous.py:
```diff
@@ -26,6 +36,86 @@ class UnitContinuous(Continuous):
     def __init__(self, transform=transforms.logodds, *args, **kwargs):
         super(UnitContinuous, self).__init__(transform=transform, *args, **kwargs)
 
+def get_named_nodes(graph):
```
My intuition is that it would be better to get this information from `model`. Is there a reason to do it this way?

jsalvatier commented on pymc3/distributions/continuous.py:
```diff
@@ -26,6 +36,86 @@ class UnitContinuous(Continuous):
+    return nodes
+
+
+def draw_values(items, point=None):
```
Can we add a use example here? I'm not actually clear on what this is supposed to do

mwibrow commented:
@jsalvatier in  #442 it was suggested that the parameters could be fixed prior to drawing `random` samples using a point (see [here](https://github.com/pymc-devs/pymc3/pull/442#issuecomment-32690342) for the description).

The `get_named_nodes` function is supposed to assist handling the specific case where a named parameter is fixed using the `point` but the distribution parameter is a `TensorVariable` instance which contains the named parameter as a node. This can occur with the `tau` and `sd` parameters in the `Normal` distribution:

``` python
with Model() as model:
    mu = Normal('mu', mu=0., tau=1.e-3)
    sigma = Gamma('sigma', alpha=1., beta=1., transform=None)
    n = Normal('n', mu=mu, sd=sigma)
```

In this case, the `tau` parameter is a `TensorVariable` instance which contains the "sigma" FreeRV. 

``` python
theano.printing.debugprint(n1.distribution.tau)

    Elemwise{mul,no_inplace} [@A] ''   
    |TensorConstant{1.0} [@B]
    |Elemwise{pow,no_inplace} [@C] ''   
      |sigma [@D]
      |TensorConstant{-2.0} [@E]
```

As far as I can see, the only way to get a value for the `tau` parameter (to generate random numbers) in this case is to evaluate the `TensorVariable` using a `theano.function` and replace sigma using the `givens` keyword. The `get_named_nodes` function seemed a reasonably efficient way to to get the named nodes for a particular `TensorVariable`.

``` python
print get_named_nodes(n1.distribution.tau)
{'sigma': sigma}
```

It's a pretty trivial example, but things could easily get more complex (sigma might not be in the point, it might need sampling via its `random` method). If this functionality is available elsewhere then that's fine. But I couldn't see it anywhere so I implemented it. 

I should point out that as things stand my version of the `draw_values` function is broken, so needs more work, but I think the `get_named_nodes` function (or some variant of it) is needed to implemented the intended functionality in the `draw_values` function. Nevertheless, I'm more than happy to ditch stuff if there is a simpler way of doing things or it is considered unnecessary.

Commit:
Rebased and merged code for random methods

Commit:
Moved get_named_nodes to model.py

Commit:
(Sort of) fixed draw_values for non-float values in point.

Commit:
Fixed ExGaussian random and Added tests for random methods.

twiecki commented:
So what's the reason to have `draw_values`? What does it sample from if I don't supply that value?

I think the main use case for this will be posterior predictive checks. For that, we could have a method that takes as input a trace, then takes a random point from that trace, sets the model to it, and then draws random samples from the observeds.

mwibrow commented:
The reason to have `draw_values` is entirely due to it's presence in the #442 PR as discussed [here](https://github.com/pymc-devs/pymc3/pull/442#issuecomment-32690342) which describes the intended functionality (as well as the doc string for the function in e64e147d4076f40bc22fb83edfb176251a68d3ed).

If a parameter (i.e., a prior) is not passed in the point to the `random` method, then it is sampled from its own random method. If the parameter is a scalar it is just evaluated (it wouldn't be in the point anyway). This I think is what was intended in the `draw_values` function  but I enhanced the definition to handle cases (i.e., un-named tensors with antecedent nodes which are *RVs) which I didn't think the original function covered. 

If `draw_values` isn't needed then that's fine. It needs a minor fix, but there's no point doing that it the whole thing is no longer necessary.

fonnesbeck commented:
Posterior predictive checks, and also methods such as approximate Bayesian computing that sample from the priors at every iteration.

twiecki commented:
@fonnesbeck But for PPC you'd sample from the posterior, I can't think of a case where one would want to sample from the prior.

fonnesbeck commented:
ABC samples from the priors at every iteration.

twiecki commented:
Hm, the ABC samplers I know don't require sampling from the prior but instead just propose moves like MH, then sample data, and have a distance function that compares the simulated data to the real data.

twiecki commented:
@fonnesbeck I'm in favor of parking that function elsewhere until we have a solid need for it. Currently we don't have support for ABC samplers and I'm hesitant to include complexity that we might or might not need one day.

@mwibrow What's your opinion on this?

twiecki commented:
Here's an example NB that shows that PPC for arbitrary models becomes trivial:
https://gist.github.com/twiecki/dca044bf16b26ea4aa1f

The function in there can easily be integrated into `pymc3`.

mwibrow commented:
@twecki if the `random` methods use numpy/scipy and if they are to be used to generate samples from an *RV instance, I think there needs to be _some_ function that provides a consistent way set a parameter to: 
1.  a value from a point (like the  PPC example), 
2. the evaluation of a theano graph with antecedent nodes which are *RVs like the `sd` or `tau` parameters from the Normal distribution (also needed in the PPC example),
3. a scalar value, which could be a theano variable - like the `nu` parameter in the T distribution (not an issue for PPC if observed RVs are guaranteed have parameters which are RVs in the model)
4. "something else" if distribution parameters are *RV instances which are not named in the point (not an issue, based on the PPC example)

So, unless I'm missing something pretty basic (or misunderstood what everyone is saying) it seems to me that it is not a question of whether the `draw_values` should be there (as it handles cases 1 & 2) but whether cases 3 & 4 are needed and if so, what `draw_values` should do for the "something else" in case 4.

Unless there is some restriction imposed on the use of the `random` methods that makes cases 3 & 4 irrelevant.

twiecki commented:
Yes @mwibrow, you're completely correct. After writing the PPC code I realized that `draw_values` is required in any case. I think drawing from the prior is probably a good default for case 4. I think I'm convinced that this is the way to go.

From my end this looks good to merge as there are even unittests now. Is everyone OK with that?

Commit:
Point is now passed to parameter random method when sampling.

twiecki commented:
Sweet, this can easily be extended to prediction, see at the end for an example using logistic regression predicting labels on held-out data: https://gist.github.com/twiecki/dca044bf16b26ea4aa1f

Going over this I noticed that random() methods are still missing for all the discrete distributions. This is pretty easy to add. Also, the `draw_value()` stuff should probably live somewhere else. Maybe in `dist_math`?

Commit:
Added random for method for flat and bounded distributions.

Commit:
Added random methods for most discrete distritbutions.

mwibrow commented:
I've had a crack at the discrete distributions. Some of the tests are a bit flaky (the `NegativeBinomial` test never passes so I've commented it out - possibly I've done some thing wrong).

Also I haven't done/finished  a `random` method for the`ZeroInflatedPoisson` as I don't recognize the parametrization that's been used. Is the `z` parameter a probability?

twiecki commented:
Not sure about `ZeroInflatedPoisson` either, I say we punt on that. Is this ready to merge from your end or is there more you think you want to do?

mwibrow commented:
@twiecki as all the checks have passed, I think I can stop fiddling for now, so its ready to merge.

fonnesbeck commented on pymc3/distributions/continuous.py:
```diff
@@ -113,6 +129,9 @@ def __init__(self, *args, **kwargs):
         self.median = 0
 
+    def random(self, point=None, size=None):
+        return np.zeros(size)
```
An uninformative distribution should not return zeros as random draws -- should be a random number on the real line (at least in principle).

aflaxman replied:
Or, perhaps, a raise a `ValueError`, since it is an improper prior.

jsalvatier replied:
Yeah, value error seems good to me.

On Sun, Jul 26, 2015 at 10:02 PM Abraham Flaxman notifications@github.com
wrote:

> In pymc3/distributions/continuous.py
> https://github.com/pymc-devs/pymc3/pull/784#discussion_r35505801:
> 
> > @@ -113,6 +129,9 @@ def **init**(self, _args, *_kwargs):
> >          super(Flat, self).**init**(_args, *_kwargs)
> >          self.median = 0
> > -    def random(self, point=None, size=None):
> > -        return np.zeros(size)
> 
> Or, perhaps, a raise a ValueError, since it is an improper prior.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pymc-devs/pymc3/pull/784/files#r35505801.

Commit:
Sampling from flat distribution now returns ValueError.

fonnesbeck commented:
This also needs a good example. I will try to come up with one.

fonnesbeck commented:
@mwibrow the `z` in `ZeroInflatedPoisson` are indicators for group membership -- see the `switch` statement in the `logp` method. They are typically modeled as a Bernoulli in a hierarchical model.

twiecki commented:
@fonnesbeck What about using some of thsi: https://gist.github.com/twiecki/dca044bf16b26ea4aa1f

fonnesbeck commented:
I notice that `random` does not pay attention to the shape of the stochastic. For example:

```
x = Normal('x', 0, 1, shape=3)
```

which specifies a vector of normals returns an array of length 1 when `random` is called:

```
In [4]: x.random()
Out[4]: array([ 0.05992264])
```

This is the subtle difference between "shape" and "size", which comes up when trying to deal with vectors of multivariate stochastics in PyMC. Should we not expect one sample (size=1) of this vector-valued variable (shape=3)?

mwibrow commented:
@fonnesbeck that makes sense. I guess something like

``` python
size = np.append(np.atleast_1d(size or 1), self.shape)
```

should do it. So if the shape is `(2,3)` and the size is `4` then the output has shape `(4, 2, 3)`

If that's acceptable, I'll sort that out in a bit.

mwibrow commented:
Actually

``` python
size = self.shape if size is None else np.append(np.atleast_1d(size), self.shape)
```

might be better, so that if `size` is not passed to the `random` method, an array of shape `self.shape` is returned. The question is whether that should also happen if `size=1`.

twiecki commented:
@mwibrow Wouldn't this remove the possibility of sampling arbitrary number of samples? I think there definitely exists a use-case for that. That almost sounds like we need to args. One, maybe `repeat` that samples `repeat * self.shape` and maybe `size` that just samples directly the shape required. If neither is passed we default to `self.shape`.

Commit:
Added get_sample_shape method for random methods.

mwibrow commented:
I've added a `get_sample_shape` which handles the size/shape problem for all random methods. I've changed the random method signature to `*args, **kwargs`. So, now
- `rv.random()` returns `self.shape` samples (or 1 sample when `self.shape` is [])
- `rv.random(x)` is the same as `rv.random(size=x)`
- `rv.random(size=x)` returns a array of samples with shape `x`
- `rv.random(repeat=x)` returns an array of samples with shape `np.append(x, self.shape)`

The `point` functionality is the same as before but should always be passed as a keyword argument.

I think that covers the use cases discussed so far.

twiecki commented:
That looks great to me.

twiecki commented:
@jsalvatier @fonnesbeck are we good to merge?

mwibrow commented:
I've just noticed that the random stuff is currently broken for some distributions when the parameter shape is not `(1,)` (e.g., for observed RVs in regression).

So a bit more work is needed before merging.

twiecki commented:
@mwibrow OK, let us know when you think it's ready.

Commit:
New random sampling infrastructure

twiecki commented:
@mwibrow For which distributions is it broken?

mwibrow commented:
@twiecki Binomial, for one, but probably any distribution which doesn't use only the `loc` and `scale` parameters of the `scipy` `rvs` methods.

Basically, it seems there are differences in the way `scipy.stats` return random samples usually depending on whether the distributions use the `loc` and `scale` parameters from the relevant scipy distribution, particularly when the shape of parameters is not `(1,)`.

For example, `scipy.stats.norm.rvs` can take any `size` which is broadcastable with the shape of the `loc` and `scale` parameters. But `scipy.stats.binom.rvs` has to have a `size` parameter which describes a shape that is exactly the same shape as the parameter(s) with the largest shape, while `numpy.random.choice` doesn't seem to care what the `size` parameter is.

As `get_sample_shape` just returns a shape to be passed as the `size` parameter to the `scipy` stuff, it doesn't take this into account so either a `ValueError` can get thrown up or the shape of the samples is unexpected.

I think I've sorted it (but it ain't pretty); I'm working through testing the shape of random samples from every distribution.

I also notice for observed RVs the shape of the distribution is independent of the RV.shape:

``` python
In [14]: import numpy as np
In [15]: from pymc3 import Model, Normal
In [16]: with Model():
    rv = Normal('n', mu=np.zeros(10), tau=np.ones(10), observed=np.ones((10,10)), shape=(3,3))
    print(rv.shape.eval(), rv.distribution.shape)
   ....:     
[10 10] [3 3]
```

Given that the `random` methods only know the distribution shape, is this going to be an issue somewhere? Ignoring the actual values, what would the expected shape of the random samples in this case?

Commit:
Updates to random stuff

Commit:
updates to random methods

twiecki commented:
Sounds like quite a mess with `numpy` and `scipy` random methods and how they chose to interpret shape parameters. There probably is no clean way to go about that...

It's pretty odd that the shape argument is accepted as the two seem to not be compatible. One more reason to remove the shape argument and only allow broadcasting. So when you say the random methods only know the distribution shape, do you mean `[3 3]` in that case?

mwibrow commented:
Yep, I mean the `shape` parameter that is given to the constructor (so `[3 3]` in the example). 

In a sense, `scipy` and `numpy` are correct in that where appropriate, they treat distributions as location and/or scale families. The problem (or least my problem) is then "massaging" the output of the different types of `rvs` method to produce consistent shapes for the `pymc3` stuff.

twiecki commented:
OK, I think it's fine to use the `shape` parameter given to the constructor for now. Detecting cases where the two don't match should be handled upstream of this PR.

Commit:
Finalised random stuff.

twiecki commented:
@mwibrow Is there anything you still thinks needs to be done or we can we merge this?

Commit:
Started multivariate random methods.

Commit:
Updated tests for multivate distributions.

mwibrow commented:
With some fingers crossed, and assuming the checks pass, this I think takes care of most things. Any glaring omissions are noted using `SkipTest`. 

The multivariate testing is incomplete as I think it depends on how issue #535 is resolved.

I also have been using `distribution.py` as a kind of dumping ground for functions which are used by the random methods. I wonder whether there should be a `dist_utils.py` module for this stuff.

But I'll stick my neck out and say it is ready to merge. At least until someone finds another really trivial example that breaks everything ;)

Commit:
Removed repeat argument to random methods

mwibrow commented:
It just occurred to me that the `repeat` argument was actually unnecessary. When the effective shape of a single sample is not `(1,)` (e.g., when broadcastable parameters are vectors) the `size` argument did the same thing as the `repeat` argument, and when the effective shape of a single sample is `(1,)`, `repeat` was a bit redundant. So, I've removed it in the latest commit.

twiecki commented:
Merged with https://github.com/pymc-devs/pymc3/commit/80249395c2e57b962231c30dd8346b1cc90eb7a8!

This was a big one, thank you so much for the hard work! I'll add an example soon.

twiecki commented:
Add docs: http://pymc-devs.github.io/pymc3/posterior_predictive/
https://github.com/pymc-devs/pymc3/blob/master/pymc3/examples/posterior_predictive.ipynb

23pointsNorth commented:
The docs don't seems to link anywhere?

dmarx commented:
@23pointsNorth pretty sure this is the notebook that link was supposed to point to: https://github.com/pymc-devs/pymc3/blob/master/docs/source/notebooks/posterior_predictive.ipynb
